File: rust_version/src/core/reactor.rs
use std::collections::HashMap;
use std::io::{Error, Result};
use std::os::unix::io::RawFd;
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;

const MAX_EVENTS : usize = 10;

pub struct Reactor{
    epoll_fd : RawFd,
    running: Arc<AtomicBool>,
    handlers: HashMap<RawFd, Box<dyn FnMut(u32) + Send>>,
}

impl Reactor{
    pub fn new() -> Result<Self> {
        let epoll_fd = unsafe { libc ::epoll_create1(0)};
        if epoll_fd == -1 {
            return Err(Error::last_os_error());
        }

        Ok(
            Reactor { 
            epoll_fd, 
            running: Arc::new(AtomicBool::new(false)), 
            handlers:HashMap::new() 
            }
        )
    }

    pub fn add_handler<F>(&mut self, fd: RawFd, events: u32, handler: F) -> Result<()>
    where
        F: FnMut(u32) + Send + 'static,
    {
        let mut ev = libc::epoll_event{
            events : events as u32,
            u64: fd as u64,
        };

        let result = unsafe{
            libc::epoll_ctl(
                self.epoll_fd,
                libc::EPOLL_CTL_ADD,
                fd,
                &mut ev as *mut libc::epoll_event,
            )
        };

        if result == -1 {
            return Err(Error::last_os_error());
        }
        
        self.handlers.insert(fd, Box::new(handler));

        Ok(())
    }

    pub fn remove_handler(&mut self, fd: RawFd) -> Result<()> {
        if !self.handlers.contains_key(&fd){
            return Ok(());
        }

        let result = unsafe {
            libc::epoll_ctl(
                self.epoll_fd,
                libc::EPOLL_CTL_DEL,
                fd,
                std::ptr::null_mut()
            )
        };

        match result {
            -1 => {
                let err = Error::last_os_error();
                match err.raw_os_error() {
                    Some(libc::EBADF) => {
                        self.handlers.remove(&fd);
                        Ok(())
                    }

                    Some(libc::ENOENT) => {
                        eprintln!("Warning: File descriptor {} not found in epoll instance", fd);
                        self.handlers.remove(&fd);
                        Ok(())
                    }
                    _ => Err(err)
                }
            }
            _ => {
                self.handlers.remove(&fd);
                Ok(())
            }
        }
    }

    pub fn run(&mut self) -> Result<()> {
        let mut events = vec![
            libc::epoll_event { events: 0, u64: 0 };
            MAX_EVENTS
        ];

        self.running.store(true, Ordering::SeqCst);

        while self.running.load(Ordering::SeqCst) {
            let nfds = unsafe {
                libc::epoll_wait(
                    self.epoll_fd,
                    events.as_mut_ptr(),
                    MAX_EVENTS as i32,
                    100, 
                )
            };

            if nfds == -1 {
                let err = Error::last_os_error();
                if err.raw_os_error() == Some(libc::EINTR) {
                    continue;
                }
                return Err(err);
            }

            for n in 0..nfds {
                let fd = events[n as usize].u64 as RawFd;
                if let Some(handler) = self.handlers.get_mut(&fd) {
                    handler(events[n as usize].events);
                }
            }
        }
        Ok(())
    }

    pub fn stop(&self) {
        println!("Reactor stopping...");
        self.running.store(false, Ordering::SeqCst);
    }

    pub fn get_epoll_fd(&self) -> RawFd {
        self.epoll_fd
    }
    
    pub fn share_running_state(&mut self, running: Arc<AtomicBool>) {
        self.running = running;
    }  
}

impl Clone for Reactor {
    fn clone(&self) -> Self {
        let epoll_fd = unsafe { libc::epoll_create1(0) };
        if epoll_fd == -1 {
            panic!("Failed to create epoll instance in clone");
        }

        Self {
            epoll_fd,
            running: Arc::clone(&self.running),
            handlers: HashMap::new(),
        }
    }
}

impl Drop for Reactor {
    fn drop(&mut self) {
        unsafe {
            libc::close(self.epoll_fd);
        }
    }
}

File: cpp_version/src/core/reactor.cpp
#include "core/reactor.h"
#include <sys/epoll.h>
#include <unistd.h>
#include <stdexcept>
#include <iostream>

Reactor::Reactor() : running_(false), events_(MAX_EVENTS) {
    epoll_fd_ = epoll_create1(0);
    if (epoll_fd_ == -1) {
        throw std::runtime_error("Failed to create epoll file descriptor");
    }
}

Reactor::~Reactor() {
    close(epoll_fd_);
}

void Reactor::add_handler(int fd, uint32_t events, std::function<void()> handler) {
    epoll_event ev;
    ev.events = events;
    ev.data.fd = fd;

    if (epoll_ctl(epoll_fd_, EPOLL_CTL_ADD, fd, &ev) == -1) {
        throw std::runtime_error("Failed to add file descriptor to epoll");
    }

    handlers_[fd] = std::move(handler);
}

void Reactor::remove_handler(int fd) {
    auto it = handlers_.find(fd);
    if (it == handlers_.end()) {
        // 处理程序不在我们的映射中，可能已经被移除
        return;
    }

    if (epoll_ctl(epoll_fd_, EPOLL_CTL_DEL, fd, nullptr) == -1) {
        if (errno == EBADF) {
            // 文件描述符无效，可能已经被关闭
            // 我们只需要从 handlers_ 中移除它
            handlers_.erase(it);
        } else if (errno == ENOENT) {
            // 文件描述符不在 epoll 实例中
            // 这是一个意外情况，因为它在我们的 handlers_ 中
            // 我们应该记录这个情况，但仍然从 handlers_ 中移除它
            std::cerr << "Warning: File descriptor " << fd << " not found in epoll instance\n";
            handlers_.erase(it);
        } else {
            // 其他错误，可能需要更严重的处理
            throw std::runtime_error("Failed to remove file descriptor from epoll: ");
        }
    } else {
        // 成功从 epoll 实例中移除，现在从 handlers_ 中移除
        handlers_.erase(it);
    }
}

void Reactor::run() {
    running_ = true;
    while (running_) {
        int nfds = epoll_wait(epoll_fd_, events_.data(), MAX_EVENTS, -1);
        if (nfds == -1) {
            if (errno == EINTR) continue;  // 被信号中断，继续循环
            throw std::runtime_error("epoll_wait failed");
        }

        for (int n = 0; n < nfds; ++n) {
            int fd = events_[n].data.fd;
            auto it = handlers_.find(fd);
            if (it != handlers_.end()) {
                it->second();  // 调用处理函数
            }
        }
    }
}

void Reactor::stop() {
    running_ = false;
}

int Reactor::get_epoll_fd()
{
    return epoll_fd_;
}